src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -950,20 +968,647 @@ void Chttp2ServerListener::Orphan() {
+
+  channelz::ListenSocketNode* channelz_listen_socket_node() const override {
+    return channelz_listen_socket_.get();
+  }
+
+  void SetServerListenerWrapper(
+      Server::ListenerWrapper* listener_wrapper) override {
+    listener_wrapper_ = listener_wrapper;
+  }
+
+  const grpc_resolved_address* resolved_address() const override {
+    return &resolved_address_;
+  }
+
+  void SetOnDestroyDone(grpc_closure* on_destroy_done) override;
+
+  void Orphan() override;
+
+ private:
+  friend class experimental::PassiveListenerImpl;
+
+  class NewActiveConnection : public LogicalConnection {
+   public:
+    class HandshakingState : public InternallyRefCounted<HandshakingState> {
+     public:
+      HandshakingState(RefCountedPtr<NewActiveConnection> connection_ref,
+                       grpc_pollset* accepting_pollset, AcceptorPtr acceptor,
+                       const ChannelArgs& args,
+                       OrphanablePtr<grpc_endpoint> endpoint);
+
+      ~HandshakingState() override;
+
+      void Orphan() override;
+
+      void StartLocked(const ChannelArgs& args);
+
+      void ShutdownLocked(absl::Status status);
+
+      // Needed to be able to grab an external weak ref in
+      // NewActiveConnection::Start()
+      using InternallyRefCounted<HandshakingState>::Ref;
+
+     private:
+      void OnTimeout();
+      static void OnReceiveSettings(void* arg, grpc_error_handle /* error */);
+      void OnHandshakeDoneLocked(absl::StatusOr<HandshakerArgs*> result);
+
+      RefCountedPtr<NewActiveConnection> const connection_;
+      grpc_pollset* const accepting_pollset_;
+      const AcceptorPtr acceptor_;
+      grpc_pollset_set* const interested_parties_;
+      Timestamp const deadline_;
+      OrphanablePtr<grpc_endpoint> endpoint_;
+      // Following fields are protected by WorkSerializer.
+      RefCountedPtr<HandshakeManager> handshake_mgr_;
+      // State for enforcing handshake timeout on receiving HTTP/2 settings.
+      absl::optional<EventEngine::TaskHandle> timer_handle_;
+      grpc_closure on_receive_settings_;
+    };
+
+    NewActiveConnection(RefCountedPtr<NewChttp2ServerListener> listener,
+                        grpc_pollset* accepting_pollset, AcceptorPtr acceptor,
+                        const ChannelArgs& args, MemoryOwner memory_owner,
+                        OrphanablePtr<grpc_endpoint> endpoint);
+    void Start(const ChannelArgs& args);
+
+    void Orphan() override;
+
+    // Needed to be able to grab an external weak ref in
+    // NewChttp2ServerListener::OnAccept()
+    using InternallyRefCounted<LogicalConnection>::RefAsSubclass;
+
+   private:
+    bool SendGoAwayImplLocked() override;
+    void DisconnectImmediatelyImplLocked() override;
+    static void OnClose(void* arg, grpc_error_handle error);
+    void OnDrainGraceTimeExpiry() ABSL_LOCKS_EXCLUDED(&mu_);
+
+    RefCountedPtr<NewChttp2ServerListener> const listener_;
```

Comment:
I'm wondering why `ActiveConnection` needs to hold a ref to `NewChttp2ServerListener`.  It looks like this is used to access the following fields in the latter:

- `memory_quota_`
- `connection_quota_`
- `tcp_server_`
- `listener_wrapper_`

If you take my suggestion elsewhere to move `memory_quota_` and `connection_quota_` into `ListenerWrapper`, then that would eliminate the need to access `NewChttp2ServerListener` for the first two.  And I think the need for `tcp_server_` will go away in the not-too-distant future when we migrate to EventEngine.

Given that, I suggest having `ActiveConnection` store pointers to `ListenerWrapper` and `grpc_tcp_server` directly.  This slightly increases memory usage in the short term, although that will go away with the EventEngine migration.  And it has the nice benefit that we can be sure that no events in `ActiveConnection` will wind up acquiring the mutex in `NewChttp2ServerListener`, thus eliminating the possibility of deadlock.